File: services/memory-router/cartographer/src/crucible_cartographer/stack_detect.py
"""Stack detection — auto-classify a repo into one of the 12 stacks.

The detector reads package manifests + filesystem-shape signals; it
deliberately doesn't run any LLM-class inference (the result drives
global_defaults selection so it must be deterministic + fast).
"""

from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass, field


@dataclass
class StackDetection:
    primary: str = ""
    secondary: list[str] = field(default_factory=list)
    versions: dict[str, str] = field(default_factory=dict)
    confidence: float = 0.0


_SIGNALS = {
    # filename → (stack id, base_confidence_bump)
    "next.config.js":            ("nextjs", 0.4),
    "next.config.ts":            ("nextjs", 0.4),
    "next.config.mjs":           ("nextjs", 0.4),
    "manage.py":                 ("django", 0.5),
    "pyproject.toml":            ("python_generic", 0.05),
    "fastapi.toml":              ("fastapi", 0.5),
    "Gemfile":                   ("rails", 0.3),
    "config.ru":                 ("rails", 0.2),
    "pom.xml":                   ("spring_boot", 0.25),
    "build.gradle":              ("spring_boot", 0.25),
    "go.mod":                    ("go_services", 0.4),
    "Cargo.toml":                ("rust_services", 0.4),
    "mix.exs":                   ("phoenix_elixir", 0.5),
    "vue.config.js":             ("vue", 0.4),
    "nuxt.config.ts":            ("vue", 0.4),
    "artisan":                   ("laravel", 0.5),
}


def detect(repo_root: str) -> StackDetection:
    """Walk the top-level repo + first-level subdirs and accumulate
    signal weights per stack id. Highest stack id with confidence >= 0.4
    becomes ``primary``; anything else above 0.2 becomes ``secondary``.
    """
    scores: dict[str, float] = {}
    versions: dict[str, str] = {}

    def _bump(stack: str, by: float) -> None:
        scores[stack] = scores.get(stack, 0.0) + by

    if not os.path.isdir(repo_root):
        return StackDetection()

    # Top-level files
    for name in os.listdir(repo_root):
        if name in _SIGNALS:
            stack, bump = _SIGNALS[name]
            _bump(stack, bump)

    # package.json is overloaded — peek to disambiguate
    pj_path = os.path.join(repo_root, "package.json")
    if os.path.isfile(pj_path):
        try:
            with open(pj_path, "r", encoding="utf-8") as f:
                pj = json.load(f)
            deps = {**pj.get("dependencies", {}), **pj.get("devDependencies", {})}
            if "next" in deps:
                _bump("nextjs", 0.45)
                versions["next"] = _parse_version(deps["next"])
            if "express" in deps:
                _bump("express", 0.45)
            if "vue" in deps:
                _bump("vue", 0.40)
        except (OSError, json.JSONDecodeError):
            pass

    # FastAPI signal lives in requirements / pyproject
    py_paths = [
        os.path.join(repo_root, "requirements.txt"),
        os.path.join(repo_root, "pyproject.toml"),
    ]
    for path in py_paths:
        if os.path.isfile(path):
            try:
                with open(path, "r", encoding="utf-8", errors="replace") as f:
                    body = f.read().lower()
            except OSError:
                continue
            if "fastapi" in body:
                _bump("fastapi", 0.45)
            if "django" in body:
                _bump("django", 0.45)
            if "flask" in body:
                _bump("flask", 0.45)

    # Rails signal — Gemfile content
    gemfile = os.path.join(repo_root, "Gemfile")
    if os.path.isfile(gemfile):
        try:
            with open(gemfile, "r", encoding="utf-8", errors="replace") as f:
                if "rails" in f.read().lower():
                    _bump("rails", 0.5)
        except OSError:
            pass

    if not scores:
        return StackDetection()

    # Resolve primary + secondary.
    sorted_stacks = sorted(scores.items(), key=lambda kv: kv[1], reverse=True)
    primary, primary_score = sorted_stacks[0]
    if primary == "python_generic":
        # python_generic by itself is too vague; demote.
        if len(sorted_stacks) > 1:
            primary, primary_score = sorted_stacks[1]
        else:
            return StackDetection()
    secondary = [s for s, sc in sorted_stacks[1:] if sc >= 0.2 and s != primary]
    return StackDetection(
        primary=primary,
        secondary=secondary,
        versions=versions,
        confidence=min(primary_score, 1.0),
    )


def _parse_version(spec: str) -> str:
    m = re.search(r"\d+(?:\.\d+){1,2}", spec)
    return m.group(0) if m else spec

File: apps/cartographer/internal/stackdetect/stackdetect.go
// Package stackdetect classifies the primary + secondary stacks of a
// repo from its file mix and manifest files.
package stackdetect

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

// Stack is the result.
type Stack struct {
	Primary   string
	Secondary []string
}

// Detect inspects manifests and file extensions.
func Detect(root string) Stack {
	hits := map[string]int{}
	tally := func(name string, n int) { hits[name] += n }

	if exists(filepath.Join(root, "package.json")) {
		// Distinguish Next/Vue/Express/etc by reading dependencies.
		body, _ := os.ReadFile(filepath.Join(root, "package.json"))
		var pj struct {
			Dependencies    map[string]string `json:"dependencies"`
			DevDependencies map[string]string `json:"devDependencies"`
		}
		_ = json.Unmarshal(body, &pj)
		merged := map[string]string{}
		for k, v := range pj.Dependencies {
			merged[k] = v
		}
		for k, v := range pj.DevDependencies {
			merged[k] = v
		}
		switch {
		case has(merged, "next"):
			tally("nextjs", 5)
		case has(merged, "nuxt"):
			tally("nuxt", 5)
		case has(merged, "vue"):
			tally("vue", 5)
		case has(merged, "@angular/core"):
			tally("angular", 5)
		case has(merged, "express"), has(merged, "fastify"), has(merged, "koa"):
			tally("express", 4)
		default:
			tally("javascript", 2)
		}
	}
	if exists(filepath.Join(root, "pyproject.toml")) || exists(filepath.Join(root, "requirements.txt")) {
		py, _ := os.ReadFile(filepath.Join(root, "pyproject.toml"))
		req, _ := os.ReadFile(filepath.Join(root, "requirements.txt"))
		all := strings.ToLower(string(py) + "\n" + string(req))
		switch {
		case strings.Contains(all, "django"):
			tally("django", 5)
		case strings.Contains(all, "fastapi"):
			tally("fastapi", 5)
		case strings.Contains(all, "flask"):
			tally("flask", 5)
		default:
			tally("python", 2)
		}
	}
	if exists(filepath.Join(root, "Gemfile")) {
		body, _ := os.ReadFile(filepath.Join(root, "Gemfile"))
		if strings.Contains(strings.ToLower(string(body)), "rails") {
			tally("rails", 5)
		} else {
			tally("ruby", 2)
		}
	}
	if exists(filepath.Join(root, "go.mod")) {
		tally("go-services", 4)
	}
	if exists(filepath.Join(root, "Cargo.toml")) {
		tally("rust-services", 4)
	}
	if exists(filepath.Join(root, "build.gradle.kts")) || exists(filepath.Join(root, "build.gradle")) ||
		exists(filepath.Join(root, "pom.xml")) {
		tally("spring-boot", 4)
	}
	if exists(filepath.Join(root, "mix.exs")) {
		tally("phoenix", 4)
	}
	if exists(filepath.Join(root, "Package.swift")) {
		tally("swift", 4)
	}
	if exists(filepath.Join(root, "composer.json")) {
		body, _ := os.ReadFile(filepath.Join(root, "composer.json"))
		if strings.Contains(strings.ToLower(string(body)), "laravel") {
			tally("laravel", 5)
		} else {
			tally("php", 2)
		}
	}

	// Pick top.
	type kv struct {
		k string
		v int
	}
	var rows []kv
	for k, v := range hits {
		rows = append(rows, kv{k, v})
	}
	for i := 0; i < len(rows); i++ {
		for j := i + 1; j < len(rows); j++ {
			if rows[j].v > rows[i].v {
				rows[i], rows[j] = rows[j], rows[i]
			}
		}
	}
	out := Stack{}
	if len(rows) > 0 {
		out.Primary = rows[0].k
		for _, r := range rows[1:] {
			out.Secondary = append(out.Secondary, r.k)
		}
	}
	return out
}

func exists(p string) bool {
	if _, err := os.Stat(p); err == nil {
		return true
	}
	return false
}

func has(m map[string]string, key string) bool {
	_, ok := m[key]
	return ok
}
